### PR TITLE
fix: adding check for blocked addresses before escrowing fees

### PR DESCRIPTION
**Pull request by seantking**
## Description

closes: #1889 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: testing/simapp/app.go
Comment (seantking) [inline testing/simapp/app.go:677]: This is pretty hacky I just left it in for discussion. I'm guessing we don't want to unblock the distribution module?

We have the following [test case](https://github.com/cosmos/ibc-go/blob/main/modules/apps/29-fee/keeper/msg_server_test.go#L188). We ensure a non-blocked module account can escrow fees. 

When I try to use the mock module account we get an AuthKeeper error saying the account doesn't exist. I'm not sure why exactly. I've left the test case as is using the distribution module account for now. 

cc: @colin-axner

**Comment by seantking** [comment]
> # [Codecov](https://codecov.io/gh/cosmos/ibc-go/pull/1890?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos) Report
> > Merging [#1890](https://codecov.io/gh/cosmos/ibc-go/pull/1890?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos) ([8525e4e](https://github.com/cosmos/ibc-go/commit/8525e4e572b91d565faef464890895e50c8183d3)) into [main](https://codecov.io/gh/cosmos/ibc-go/commit/af4e6516b7826e7b8fcb062a3ea4608b4bdaf532?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos) ([af4e651](https://github.com/cosmos/ibc-go/commit/af4e6516b7826e7b8fcb062a3ea4608b4bdaf532)) will **decrease** coverage by `0.12%`.
> > The diff coverage is `68.80%`.
> 
> [Impacted file tree graph](https://codecov.io/gh/cosmos/ibc-go/pull/1890?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos)
> 
> ```diff
> @@            Coverage Diff             @@
> ##             main    #1890      +/-   ##
> ==========================================
> - Coverage   80.04%   79.92%   -0.13%     
> ==========================================
>   Files         166      166              
>   Lines       12421    12446      +25     
> ==========================================
> + Hits         9943     9947       +4     
> - Misses       2013     2030      +17     
> - Partials      465      469       +4     
> ```
> 
> [Impacted Files](https://codecov.io/gh/cosmos/ibc-go/pull/1890?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos)	Coverage Δ	
> [cmd/build_test_matrix/main.go](https://codecov.io/gh/cosmos/ibc-go/pull/1890/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos#diff-Y21kL2J1aWxkX3Rlc3RfbWF0cml4L21haW4uZ28=)	`70.83% <0.00%> (-3.46%)`	
> [...27-interchain-accounts/controller/keeper/keeper.go](https://codecov.io/gh/cosmos/ibc-go/pull/1890/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos#diff-bW9kdWxlcy9hcHBzLzI3LWludGVyY2hhaW4tYWNjb3VudHMvY29udHJvbGxlci9rZWVwZXIva2VlcGVyLmdv)	`94.73% <ø> (ø)`	
> [.../apps/27-interchain-accounts/host/keeper/keeper.go](https://codecov.io/gh/cosmos/ibc-go/pull/1890/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos#diff-bW9kdWxlcy9hcHBzLzI3LWludGVyY2hhaW4tYWNjb3VudHMvaG9zdC9rZWVwZXIva2VlcGVyLmdv)	`83.33% <ø> (ø)`	
> [modules/apps/27-interchain-accounts/module.go](https://codecov.io/gh/cosmos/ibc-go/pull/1890/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos#diff-bW9kdWxlcy9hcHBzLzI3LWludGVyY2hhaW4tYWNjb3VudHMvbW9kdWxlLmdv)	`55.55% <0.00%> (-2.14%)`	
> [modules/apps/29-fee/keeper/keeper.go](https://codecov.io/gh/cosmos/ibc-go/pull/1890/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos#diff-bW9kdWxlcy9hcHBzLzI5LWZlZS9rZWVwZXIva2VlcGVyLmdv)	`92.48% <ø> (ø)`	
> [modules/apps/29-fee/module.go](https://codecov.io/gh/cosmos/ibc-go/pull/1890/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos#diff-bW9kdWxlcy9hcHBzLzI5LWZlZS9tb2R1bGUuZ28=)	`54.54% <0.00%> (-1.02%)`	
> [modules/apps/transfer/keeper/keeper.go](https://codecov.io/gh/cosmos/ibc-go/pull/1890/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos#diff-bW9kdWxlcy9hcHBzL3RyYW5zZmVyL2tlZXBlci9rZWVwZXIuZ28=)	`91.17% <ø> (ø)`	
> [modules/apps/transfer/keeper/migrations.go](https://codecov.io/gh/cosmos/ibc-go/pull/1890/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos#diff-bW9kdWxlcy9hcHBzL3RyYW5zZmVyL2tlZXBlci9taWdyYXRpb25zLmdv)	`93.10% <ø> (-0.23%)`	
> [modules/apps/transfer/keeper/relay.go](https://codecov.io/gh/cosmos/ibc-go/pull/1890/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos#diff-bW9kdWxlcy9hcHBzL3RyYW5zZmVyL2tlZXBlci9yZWxheS5nbw==)	`88.11% <0.00%> (ø)`	
> [modules/apps/transfer/module.go](https://codecov.io/gh/cosmos/ibc-go/pull/1890/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos#diff-bW9kdWxlcy9hcHBzL3RyYW5zZmVyL21vZHVsZS5nbw==)	`57.14% <0.00%> (-0.93%)`	
> ... and [46 more](https://codecov.io/gh/cosmos/ibc-go/pull/1890/diff?src=pr&el=tree-more&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=cosmos)

Not too sure what's going on with the code cov here. I don't have access to the website either.

**Review by crodriguezvega — APPROVED** [review]
Looks good to me. Thanks @seantking for fixing the issue so quickly! ⚡

File: testing/simapp/app.go
Comment (damiannolan) [reply testing/simapp/app.go:677]: I was looking at the test case "refund account is a module account" and trying to use `ibcmock.ModuleName` instead of `disttypes.ModuleName`:

```
"refund account is module account",
func() {
	msg.Signer = suite.chainA.GetSimApp().AccountKeeper.GetModuleAddress(disttypes.ModuleName).String()
	expPacketFee := types.NewPacketFee(fee, msg.Signer, nil)
	expFeesInEscrow = []types.PacketFee{expPacketFee}
},
true,
```

The account does not exist error is being returned from [here in escrow.go](https://github.com/cosmos/ibc-go/blob/main/modules/apps/29-fee/keeper/escrow.go#L24). And interestingly if we actually use [`AccountKeeper.GetModuleAccount()`](https://github.com/cosmos/cosmos-sdk/blob/v0.46.0/x/auth/keeper/keeper.go#L211) we will not encounter that error as it calls directly into `GetModuleAccountAndPermissions()` which [creates and sets the account in `x/auth`](https://github.com/cosmos/cosmos-sdk/blob/v0.46.0/x/auth/keeper/keeper.go#L201-L204).

This differs from [`GetModuleAddress`](https://github.com/cosmos/cosmos-sdk/blob/v0.46.0/x/auth/keeper/keeper.go#L165) which just checks the `permAddrs` map created on keeper initialisation. 

It's definitely pretty ambiguous as to what constitutes a blocked address and what doesn't. Without this logical OR added for `distrtypes.ModuleName` it is and will be in the `blockedAddrs` map in `x/bank`. However, its usage in `x/distribution` when for example, [withdrawing delegation rewards](https://github.com/cosmos/cosmos-sdk/blob/e39d84e06fafc33dcec85db12242e5f481ed0385/x/distribution/keeper/delegation.go#L172) uses `SendCoinsFromModuleToAccount` which only checks that the [recipientAddr](https://github.com/cosmos/cosmos-sdk/blob/e39d84e06fafc33dcec85db12242e5f481ed0385/x/bank/keeper/keeper.go#L322) is not a blocked address.

When I use the following in the `malleate()` func, we don't fail on the account not existing but instead fail with an "insufficient funds" error.

```
moduleAcc := suite.chainA.GetSimApp().AccountKeeper.GetModuleAccount(suite.chainA.GetContext(), ibcmock.ModuleName)
msg.Signer = moduleAcc.GetAddress().String()
expPacketFee := types.NewPacketFee(fee, msg.Signer, nil)
expFeesInEscrow = []types.PacketFee{expPacketFee}
```

It would be nice to get some more clarity around module accounts in general.

Comment (colin-axner) [reply testing/simapp/app.go:677]: If the blocked address logic gets removed from the SDK, what test case would we want to have? I'd say probably still wise to just use the mock module account since it may not make sense long term to use distr (as community pool might be removed from it)

Maybe we can add the mock module to the list of genAccs in `NewTestChainWithValSet`?

Also I think you can use `GetModuleAddress().String()`

**Review by colin-axner — APPROVED** [review]
ACK. Nice work. Would prefer to resolve the testing/simapp/app.go discussion before merging, but I'm ok doing it in a followup

File: modules/apps/29-fee/keeper/msg_server.go
Comment (colin-axner) [reply modules/apps/29-fee/keeper/msg_server.go:129]: Agreed

File: modules/apps/29-fee/keeper/msg_server.go
Comment (crodriguezvega) [inline modules/apps/29-fee/keeper/msg_server.go:86]: Maybe instead of adding a new error, do something like?

```go
if k.bankKeeper.BlockedAddr(refundAcc) {
 	return sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "%s is not allowed to escrow fees", refundAcc)
 }
```

**Comment by crodriguezvega** [comment]
@seantking @colin-axner I am wondering if we should also add a check to ensure that the payee address is not blocked in [the rpc handler for the `MsgRegisterPayee` message](https://github.com/cosmos/ibc-go/blob/main/modules/apps/29-fee/keeper/msg_server.go#L20). This is executed on the source chain with a source chain address, so we could potentially already make sure that the payee is not blocked and prevent failures on distribution. What do you think?

**Comment by colin-axner** [comment]
> I am wondering if we should also add a check to ensure that the payee address is not blocked in [the rpc handler for the MsgRegisterPayee message](https://github.com/cosmos/ibc-go/blob/main/modules/apps/29-fee/keeper/msg_server.go#L20).

Yes definitely, nice catch!

**Comment by seantking** [comment]
> @seantking @colin-axner I am wondering if we should also add a check to ensure that the payee address is not blocked in [the rpc handler for the `MsgRegisterPayee` message](https://github.com/cosmos/ibc-go/blob/main/modules/apps/29-fee/keeper/msg_server.go#L20). This is executed on the source chain with a source chain address, so we could potentially already make sure that the payee is not blocked and prevent failures on distribution. What do you think?

Great idea! Added a check and test for this.

File: modules/apps/29-fee/keeper/msg_server.go
Comment (seantking) [reply modules/apps/29-fee/keeper/msg_server.go:95]: :handshake: done

File: modules/apps/29-fee/keeper/msg_server_test.go
Comment (colin-axner) [inline modules/apps/29-fee/keeper/msg_server_test.go:8]:
```suggestion
	sdk "github.com/cosmos/cosmos-sdk/types"

	transfertypes "github.com/cosmos/ibc-go/v5/modules/apps/transfer/types"
	"github.com/cosmos/ibc-go/v5/modules/apps/29-fee/types"
	clienttypes "github.com/cosmos/ibc-go/v5/modules/core/02-client/types"
```

Comment (colin-axner) [reply modules/apps/29-fee/keeper/msg_server_test.go:8]: probably should do this locally so it is gofmt'd

Comment (seantking) [reply modules/apps/29-fee/keeper/msg_server_test.go]: done. thanks

**Review by colin-axner — APPROVED** [review]
followup ACK

**Review by crodriguezvega — APPROVED** [review]
🥇

**Pull request merged**
